### DISCUSSION TITLE
Visualise 3D voxel grids / occupancy

### DISCUSSION THREAD

**Discussion by Lapayo** [foxglove/community#669] [Q&A]
Hey!
First of all, thanks for developing foxglove studio. It really helps me with my day to day visualisation needs.
However, I currently struggle with visualizing 3D voxel grids (Occupancy).
I already tried visualizing them using point clouds of voxel centers. That works quite okay, but due to the 2D rendering of points, there is no feeling for depth and height. I also tried using SceneEntity with cuboids but then foxglove studio becomes very slow and unresponsive (e.g. 10k-50k boxes) 
Is there a better way to achieve that? 

As a reference, this would be something I am trying to achieve: 
https://github.com/wzzheng/TPVFormer/blob/main/assets/Visualize.png

I really appreciate any help!

Simon

**Comment by amacneil** [comment]
It does seem like a native voxel grid message schema would be helpful here. A similar but different request we have also heard is for a height map/grid.

The only workarounds I can think of are SceneEntity cubes (which it sounds like you tried), or with the PointCloud message you can change the "point size" setting:

<img width="310" alt="image" src="https://user-images.githubusercontent.com/637671/227732935-8286d6e5-b1fb-4b79-8140-6013166f4e72.png">

**Reply by Lapayo** [reply]
Thanks for your answer. Yes, I tried using SceneEntity. I also tried increasing the point size, but the results are not appealing. 
From your feeling: Do you think the issue is just the message schema and adding a dedicated one producing the same boxes just with a different message would solve the issue (i.e. is parsing the message the performance issue)? Or will rendering that amount of boxes still be an issue?

**Reply by defunctzombie** [reply]
Would a point shape of "box" achieve what you want using the point cloud schemas or do you need control on the size of each voxel separate from other voxels?

**Reply by fkballweg** [reply]
In my case I'm using OpenVDB Data. Here boxes do have a different size. Smallest box is called a voxel.
see tree topology here: https://www.openvdb.org/images/example_gl.jpg

**Reply by Lapayo** [reply]
> Would a point shape of "box" achieve what you want using the point cloud schemas or do you need control on the size of each voxel separate from other voxels?

Yes, for me that would be exactly what I need. I have a fixed, symmetric voxel size. Do you think that this is easy to add? Could I somehow assist?

**Reply by snosenzo** [reply]
I think this might be a bit harder than that unfortunately. If we're looking to modify the [`PointClouds.ts` scene extension](https://github.com/foxglove/studio/blob/main/packages/studio-base/src/panels/ThreeDeeRender/renderables/PointClouds.ts) then we would have to use an entirely different `InstancedBufferGeometry` (Instead of `BufferGeometry`) and `MeshStandardMaterial` (instead of `PointsMaterial`). These would both incur a higher performance overhead and I can't say that they would be any better than using the CubePrimitives in the SceneEntity. There might be a better way that I'm unaware of but  I think we'll probably want a more dedicated approach for this, especially to handle larger datasets.

**Comment by fkballweg** [comment]
Hi Simon, 
we have used the same approach as you did using SceneEntity and CubePrimitive Type. 
In our current scenario we have ~300k Boxes. We achieved an ok-ish performance loading static Voxels only once for their lifetime. 
We're are also thinking about a dedicated Voxel-Schema - but haven't figured out how to visualize a newly introduced schema in e.g. the 3D Panel.

**Reply by Lapayo** [reply]
Hi thanks for your answer. Unfortunately, my voxels aren't static. But also without playing my foxglove drops to a very low frame rate.
